Add step to compare current page to a regex

Oursin opened the pull request:
Here is a proposal for a new builtin step that checks the page relative to a given regex.

The idea is that sometimes some parameters are located in the path of a page, such as an ID (`/user/123/profile`). In those cases, using a regex is much easier than trying to assert the exact value of those IDs, especially if the ID is random.

Adding a step `Verify page is like {{regex}}` will help in those cases.

**Changes**
Added `verifyAtPageRegex` function in `packages/js/browserinstance.js` that does the match against the given regex.
Added `Verify page is like {{regex}}` step in `packages/browser.smash`
Added tests in `tests/packages/browser.smash`

faso commented:
Since the existing one is `Verify at page`, maybe this one should be `Very at page like {{regex}}`?

Oursin commented:
It does seem better, i'll change that.
I think it would also be great to add regex alternatives to the `Wait until at page {{titleOrUrl}}` step, do you have any idea for a name ?

faso commented:
That would be good for consistency I think. Let's see what @vptes1 thinks.

vptes1 commented:
This looks good! 

What do you think of merging this code with `Verify at page {{titleOrUrl}}`, such that if the title or url doesn't contain the given string, it tries parsing it as a regex? Just want to reduce the number of steps that are out there.

And I agree that we'd need corresponding `Wait until` and `Wait until...(up to {{n}} secs)` steps

Oursin commented:
Thanks, I just merged the code, which now will support `Wait until` by default, since the same JS function is called.
I also updated the tests and added cases for the `Wait until` steps.

vptes1 commented on packages/js/browserinstance.js:
```diff
@@ -725,8 +725,12 @@ class BrowserInstance {
         try {
             await this.driver.wait(async () => {
                 obj = await this.executeScript(function(titleOrUrl) {
+                    let isMatched: document.title.toLowerCase().indexOf(titleOrUrl.toLowerCase()) != -1 || window.location.href.indexOf(titleOrUrl) != -1;
```
You mean `let isMatched =`?

Oursin replied:
It was a typo I made when copy-pasting my code, sorry. It should be fixed.

vptes1 commented on packages/js/browserinstance.js:
```diff
@@ -725,8 +725,12 @@ class BrowserInstance {
+                    let isMatched = document.title.toLowerCase().indexOf(titleOrUrl.toLowerCase()) != -1 || window.location.href.indexOf(titleOrUrl) != -1;
+                    let documentMatch = document.title.match(titleOrUrl);
+                    let locationMatch = window.location.href.match(titleOrUrl);
+                    isMatched = isMatched || documentMatch && documentMatch[0] === document.title || locationMatch && locationMatch[0] === window.location.href;
```
I'd put parenthesis around `documentMatch && documentMatch[0] === document.title` and `locationMatch && locationMatch[0] === window.location.href`. 

Also, maybe rename `documentMatch` to `titleRegexMatch` and `locationMatch` to `urlRegexMatch` or something like that.

Oursin replied:
Done !

Oursin commented on tests/packages/browser.smash:
```diff
@@ -1203,19 +1218,36 @@ Open test page [
                 Wait until at page 'badstr'
                     - Verify error $s #manual
 
-        - Wait until at page (up to n secs)
```
Woops, didn't see that. They are back.
